src/rust/src/x509.rs

```diff
@@ -29,13 +29,147 @@ lazy_static::lazy_static! {
+}
+
+// DisplayText also allows BMPString, which we currently do not support.
+#[allow(clippy::enum_variant_names)]
+#[derive(asn1::Asn1Read)]
+enum DisplayText<'a> {
+    IA5String(asn1::IA5String<'a>),
+    UTF8String(asn1::Utf8String<'a>),
+    VisibleString(asn1::VisibleString<'a>),
+}
+
+fn parse_cp(py: pyo3::Python<'_>, ext_data: &[u8]) -> Result<pyo3::PyObject, PyAsn1Error> {
+    let cp = asn1::parse_single::<asn1::SequenceOf<'_, PolicyInformation<'_>>>(ext_data)?;
+    let x509_module = py.import("cryptography.x509")?;
+    let certificate_policies = pyo3::types::PyList::empty(py);
+    for policyinfo in cp {
+        let pi = x509_module
+            .call_method1(
+                "ObjectIdentifier",
+                (policyinfo.policy_identifier.to_string(),),
+            )?
+            .to_object(py);
+        let py_pqis = match policyinfo.policy_qualifiers {
+            Some(policy_qualifiers) => {
+                let py_pq = pyo3::types::PyList::empty(py);
+                for pqi in policy_qualifiers {
+                    let qualifier = match pqi.qualifier {
+                        Qualifier::CpsUri(data) => {
+                            if pqi.policy_qualifier_id == *CP_CPS_URI_OID {
+                                pyo3::types::PyString::new(py, data.as_str()).to_object(py)
+                            } else {
+                                return Err(PyAsn1Error::from(
+                                    pyo3::exceptions::PyValueError::new_err(
+                                        "CpsUri ASN.1 structure found but OID did not match",
+                                    ),
+                                ));
+                            }
+                        }
+                        Qualifier::UserNotice(un) => {
+                            if pqi.policy_qualifier_id != *CP_USER_NOTICE_OID {
+                                return Err(PyAsn1Error::from(
+                                    pyo3::exceptions::PyValueError::new_err(
+                                        "UserNotice ASN.1 structure found but OID did not match",
+                                    ),
+                                ));
+                            }
+                            let et = match un.explicit_text {
+                                Some(DisplayText::IA5String(data)) => {
+                                    pyo3::types::PyString::new(py, data.as_str()).to_object(py)
+                                }
+                                Some(DisplayText::UTF8String(data)) => {
+                                    pyo3::types::PyString::new(py, data.as_str()).to_object(py)
+                                }
+                                Some(DisplayText::VisibleString(data)) => {
+                                    pyo3::types::PyString::new(py, data.as_str()).to_object(py)
+                                }
+                                None => py.None(),
```

Comment:
Factor this into a function so that we don't need coverage for each string type in both paths